client-apis/decrypt.md

```diff
@@ -69,65 +98,259 @@ This default CMM MUST obtain the decryption materials required for decryption.
+the input [encrypted message](#encrypted-message).
 
-Decrypt MUST use the encryption algorithm obtained from the algorithm suite.  
-The cipher key used for decryption is the derived key outputted by the [KDF algorithm](../framework/algorithm-suites.md#key-derivation-algorithm)
-specified by the algorithm suite.  
-The input to the KDF algorithm is the plaintext data key.
+This output MAY be satisfied by outputting a [parsed header](#parsed-header) containing this value.
 
-The AAD used in decryption is the [Message Body AAD](../data-format/message-body-aad.md), constructed as follows:
+### Parsed Header
 
-- Message ID: This value is the same as the message ID in the parsed message header.
-- Body AAD Content: This value depends on whether the encrypted content being decrypted is within a [regular frame](../data-format/message-body.md#regular-frame) ,
-  a [final frame](../data-format/message-body.md#final-frame) or is [non framed](../data-format/message-body.md#non-framed-data).
-  Refer to [Message Body AAD](../data-format/message-body-aad.md) specification for more information.
-- Sequence Number: This value is the [sequence number](../data-format/message-body.md#sequence-number) of the frame being decrypted, if the message contains framed data.
-  If the message contains non framed data, then this value is 1.
-- Content Length: TODO
+A collection of deserialized fields of the [encrypted message's](#encrypted-message) header.
 
-If the algorithm suite has a signature algorithm, decrypt MUST verify the message footer using the specified signature algorithm,
-by using the verification key obtained from the decryption materials.
+## Operation
 
-## Security Considerations
+The Decrypt operation is divided into several distinct parts:
+
+- [Parse the header](#parse-the-header)
+- [Get the decryption materials](#get-the-decryption-materials)
+- [Verify the header](#verify-the-header)
+- [Decrypt the body](#decrypt-the-body)
+- [Verify the signature (optional)](#verify-the-signature-optional)
 
-[TODO]
+This operation MUST perform these steps in order.
+If any of these steps fails, this operation MUST fail and indicate a failure to the caller.
 
-## Apendix
+If the input encrypted message is not being streamed to this operation,
+all output MUST NOT be released until after these steps complete successfully.
+
+If the input encrypted message is being streamed to this operation:
+
+- Output MUST NOT be released until otherwise indicated.
+- If no more input plaintext bytes MAY become available and this operation
+  is unable to complete the above steps with the available plaintext,
+  this operation MUST halt and indicate a failure to the caller.
+- If this operation successfully completes the above steps
+  and has processed all available input bytes,
+  this operation MAY immediately complete and succeed,
+  or it MAY wait until an end of the input bytes is indicated
+  to complete and succeed.
```

Comment:
I don't get this one. How can it succeed if it doesn't know if more input is coming and hence hitting the case immediately below?

Reply:
Deciding to cut this wording out for now, as this is more of a question of how implementations choose to interpret streaming input. Will determine offline if there are any additional constraints we should add in the Input section above. But for now this should be covered by "there is some mechanism to determine end of input", and how each implementation defines that mechanism may differ. For Python, the mechanism is the operation itself determining that it has received a valid encrypted message, and opting to not read any more bytes.